Introduce the Variable Resolver Extension

azatsarynnyy opened the pull request:
Signed-off-by: Artem Zatsarynnyi <azatsary@redhat.com>

PR closes #1561

This PR introduces the `variable-resolver` extension which provides support of variable substitution mechanism inside strings. Extension adds:
- a contribution point to provide a hook that allows any extensions to contribute its own variables e.g.:
```typescript
export class EditorVariableContribution implements VariableContribution {
    registerVariables(variables: VariableRegistry): void {
        variables.registerVariable({
            name: 'lineNumber',
            description: 'The current line number in the current file',
            resolve: () => {
                const currentEditor = ...
                return Promise.resolve(`${currentEditor.cursor.line + 1}`);
            }
        });
    }
}
```
- `VariableResolverService` allows to resolve all known variables in a string e.g.:
```typescript
const resolved = await service.resolve('file: ${file}; line: ${lineNumber}');
expect(resolved).is.equal('file: package.json; line: 6');
```
- `VariableQuickOpenService` allows to get list of all registered variables e.g.:

![screenshot from 2018-03-26 17-20-30](https://user-images.githubusercontent.com/1636395/37914365-c03f22e4-311f-11e8-980a-7bbb02389b9f.png)

Note, this PR doesn't contribute any variables. I'm going to provide a couple of separate PRs a bit later to add support for variables in `tasks.json` and add the [VSCode's predefined variables](https://code.visualstudio.com/docs/editor/variables-reference#_predefined-variables) to add a little bit more compatibility with VSCode Tasks.

kittaakos commented on packages/variable-resolver/src/browser/variable.spec.ts:
```diff
@@ -0,0 +1,102 @@
+import { ILogger } from '@theia/core/lib/common';
+import { MockLogger } from '@theia/core/lib/common/test/mock-logger';
+import { Variable, VariableRegistry } from './variable';
+import { Disposable } from 'vscode-jsonrpc';
```
I do not know whether it was intentional or not, but we usually user `@theia/core/common/disposable`. I guess the tooling did it.

azatsarynnyy replied:
Yes, it was VSCode's work ))
Fixed import.

kittaakos commented on packages/variable-resolver/src/browser/variable.ts:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * A unique name of this variable.
+     */
+    name: string;
```
Did you omit the `readonly` on purpose or not? If by accident, please add it. Thank you!

azatsarynnyy replied:
Just overlooked. Annotated both properties as `readonly`.

kittaakos commented on packages/variable-resolver/src/browser/variable.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /**
+     * Get a variable for the given name or 'undefined' if none.
```
Could you please use `` ` `` instead of `'`? The documentation looks better in VSCode. What do you think?

azatsarynnyy replied:
Absolutely agree. Changed

kittaakos commented on packages/variable-resolver/src/browser/variable-quick-open-service.ts:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    run(mode: QuickOpenMode): boolean {
+        if (mode !== QuickOpenMode.OPEN) {
```
Then, it is always `false`, isn't it?

azatsarynnyy replied:
Sure. Changed.

kittaakos commented on packages/variable-resolver/src/browser/variable-resolver-frontend-contribution.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const LIST_VARIABLES: Command = {
+    id: 'variable.list',
+    label: 'Variable: list all'
```
Could we have `Variable: List All` instead? At least, that was the convention for instance in the Git extension so far.

azatsarynnyy replied:
Updated

epatpol commented:
Looks very good! Maybe add a short example in the `README` on how you can use the substitutions i.e implement a contribution and inject the service?

akosyakov commented on packages/variable-resolver/src/browser/variable-quick-open-service.ts:
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
```
I think the copyright should be under Apache 2 license:
```
/*
 * Copyright (C) 2018 Read Hat, Inc. and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
```
cc @svenefftinge

azatsarynnyy replied:
I've updated all files

akosyakov commented on packages/variable-resolver/src/browser/variable.ts:
```diff
@@ -0,0 +1,93 @@
+     * 'undefined' if variable cannot be resolved.
+     * Never reject.
+     */
+    resolve(): Promise<string | undefined>;
```
You can use `MaybePromise` here, that implementors does not deal with promises. You can resolve a maybe promise with `await` syntax or `Promise.resolve`.

azatsarynnyy replied:
Yeah, with `MaybePromise` it looks more flexible. I've updated it and also slightly refactored the `VariableResolverService`.
Thanks for the advice.

akosyakov commented on packages/variable-resolver/src/browser/variable-quick-open-service.ts:
```diff
@@ -0,0 +1,68 @@
+        const quickOpenItems = this.variableRegistry.getVariables().map(
+            v => new VariableQuickOpenItem(v.name, v.description)
+        );
+        this.items = [...quickOpenItems];
```
`this.items = quickOpenItems` ? `map` fn creates a new array

azatsarynnyy replied:
Sure ) Changed

azatsarynnyy commented:
@kittaakos @epatpol @akosyakov 
Guys, thanks for the reviews. I've addressed all your comments and rebased the branch.
@epatpol I've updated the README with the examples of using.

epatpol commented:
Yes I didn't expect a readme that extensive, very well done thank you!

azatsarynnyy commented:
Guys, could someone give me the approval to merge this PR if it looks good for you?

kittaakos commented:
> give me the approval

The PR looks great. @akosyakov, do you have any additional remarks on this?
